File: C++/number-of-prefix-connected-groups.cpp
// Time:  O(n * k)
// Space: O(n * k)

// freq table
class Solution {
public:
    int prefixConnected(vector<string>& words, int k) {
        unordered_map<string, int> cnt;
        for (const auto& w : words) {
            if (size(w) < k) {
                continue;
            }
            ++cnt[w.substr(0, k)];
        }
        int result = 0;
        for (const auto& [_, v] : cnt) {
            if (v >= 2) {
                ++result;
            }
        }
        return result;
    }
};

File: Python/number-of-prefix-connected-groups.py
# Time:  O(n * k)
# Space: O(n * k)

import collections


# freq table
class Solution(object):
    def prefixConnected(self, words, k):
        """
        :type words: List[str]
        :type k: int
        :rtype: int
        """
        cnt = collections.defaultdict(int)
        for w in words:
            if len(w) < k:
                continue
            cnt[w[:k]] += 1
        return sum(v >= 2 for v in cnt.itervalues())
